File: dcp-solutions/2024-09-17/solution.go
```markdown
package main

import "container/list"

type TreeNode struct {
	Val int
	Left *TreeNode
	Right *TreeNode
}

func countUnivalSubtrees(root *TreeNode) int {
	var count int
	var dfs func(*TreeNode)
	
	dfs = func(node *TreeNode) {
		if node == nil {
			return 
		}
		
		left := true
		right := true
		var l, r []int
		var stack container.List
		
		stack.PushBack(node, 0)
		
		for !stack.IsEmpty() {
			curr := stack.Front().Value.(TreeNode)
			stack.Remove(stack.Front())
			
			l = append(l, curr.Val)
			r = append(r, curr.Val)
			left &= curr.Val == curr.Left.Val
			right &= curr.Val == curr.Right.Val
            
			if curr.Left != nil {
				stack.PushBack(curr.Left, 1)
			}
			if curr.Right != nil {
				stack.PushBack(curr.Right, 0)
			}
			
		}

		if left && right {
			count++
		}
		
		for len(l) > 1 {
			mid := (len(l)-1)/2
			dfs(&TreeNode{Val: l[mid], Left: &TreeNode{Val: l[mid]}, Right: &TreeNode{Val: r[mid]}})
			l = l[:mid]
			r = r[mid+1:]
		}
		
		for len(r) > 1 {
			mid := (len(r)-1)/2
			dfs(&TreeNode{Val: l[mid], Left: &TreeNode{Val: r[mid]}, Right: &TreeNode{Val: r[mid+1]}})
			l = l[mid+1:]
			r = r[:mid]
		}
	}

	dfs(root)
	return count
}
```

File: dcp-solutions/2024-09-17/solution.py
```python
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

def countUnivalSubtrees(root):
    def check(node, val):
        if node is None:
            return True
        if node.val != val:
            return False
        return check(node.left, val) and check(node.right, val)

    def dfs(node):
        if node is None:
            return 1, 0
        same_left = int(check(node.left, node.val))
        same_right = int(check(node.right, node.val))
        return same_left + same_right + (1 if node.left is None or node.right is None else dfs(node.left)[0] + dfs(node.right)[0])

    return dfs(root)[1]
```